bitcoin/tx.c

```diff
@@ -726,3 +726,73 @@ wally_tx_output_get_amount(const struct wally_tx_output *output)
 
 	return amount;
 }
+
+/* Various weights of transaction parts. */
+size_t bitcoin_tx_core_weight(size_t num_inputs, size_t num_outputs)
+{
+	size_t weight;
+
+	/* version, input count, output count, locktime */
+	weight = (4 + varint_size(num_inputs) + varint_size(num_outputs) + 4)
+		* 4;
+
+	/* A couple of things need to change for elements: */
+	if (chainparams->is_elements) {
+                /* Each transaction has surjection and rangeproof (both empty
+		 * for us as long as we use unblinded L-BTC transactions). */
+		weight += 2 * 4;
+
+		/* An elements transaction has 1 additional output for fees */
+		weight += (8 + 1) * 4; /* Bitcoin style output */
+		weight += (32 + 1 + 1 + 1) * 4; /* Elements added fields */
+	}
+	return weight;
+}
+
+size_t bitcoin_tx_output_weight(size_t outscript_len)
+{
+	size_t weight;
+
+	/* amount, len, scriptpubkey */
+	weight = (8 + varint_size(outscript_len) + outscript_len) * 4;
+
+	if (chainparams->is_elements) {
+		/* Each output additionally has an asset_tag (1 + 32), value
+		 * is prefixed by a version (1 byte), an empty nonce (1
+		 * byte), two empty proofs (2 bytes). */
+		weight += (32 + 1 + 1 + 1) * 4;
+	}
+
+	return weight;
+}
+
+/* We grind signatures to get them down to 71 bytes (+1 for sighash flags) */
+size_t bitcoin_tx_input_sig_weight(void)
+{
+	return 1 + 71 + 1;
+}
+
+/* We only do segwit inputs, and we assume witness is sig + key  */
+size_t bitcoin_tx_simple_input_weight(bool p2sh)
+{
+	size_t weight;
+
+	/* Input weight: txid + index + sequence */
+	weight = (32 + 4 + 4) * 4;
+
+	/* We always encode the length of the script, even if empty */
+	weight += 1 * 4;
+
+	/* P2SH variants include push of <0 <20-byte-key-hash>> */
+	if (p2sh)
+		weight += 23 * 4;
```

Comment:
are reedeemscripts prefixed with a length in addition to the scriptSig length? if not, this overcounts by 1 (23 is 20 + 1 pushdata + 00 + 1 len)

Reply:
Agreed, good catch!  We already count the length above.

Reply:
I went and looked this up, `23` is actually correct, as you do prefix redeemscripts with their length (in addition to prefixing the scriptSig with it's length).

Example:

`23 22 00209c9c8e1519dacba326d9c2987e15944e9ff5d19315d4af33d2511385d66e5755`

Reply:
Yes, you're right.  Code review FTW! (I simply moved the code).